Monitoring: Add Prometheus Targets UI pages

kyoto opened the pull request:
Epic: https://issues.redhat.com/browse/MON-1079

FYI @cshinn, @thile01

![screenshot-2](https://user-images.githubusercontent.com/460802/145060564-9ff04127-4d7a-4e7d-a305-a2c82261310d.png)

![screenshot-3](https://user-images.githubusercontent.com/460802/145060629-d84abc7c-1607-405e-aa2e-1a4635756507.png)

kyoto commented:
Adding `docs-approved` because the epic already has the `doc-ack` label.

/label docs-approved

spadgett commented on frontend/public/components/monitoring/targets.tsx:
```diff
@@ -0,0 +1,365 @@
+        />
+        <div className="row">
+          <div className="col-xs-12">
+            <Table
```
Don't redo this, but in general new code should use the updated table components from #8829.

spadgett reviewed (approved):
/lgtm

kyoto commented:
/cc @lihongyan1

kyoto commented:
Adding `px-approved` because the epic already has the `px-ack` label.
FYI @RickJWagner

/label px-approved

lihongyan1 commented:
The Name filter drop down with value Name&Label should be change as Endpoint&Monitor

lihongyan1 commented:
The Name filter drop down with value Name&Label should be change as Endpoint&Monitor

lihongyan1 commented:
Order by Monitor can't works well, Not sure why some targets with null Monitor

lihongyan1 commented:
It would be better if admin can filter targets by namespace/project

kyoto commented:
Thanks @lihongyan1. I've made the following changes to address the issues you found.
- Changed the text filter to match against Endpoint and Namespace (instead of Endpoint and Monitor)
- Changed the text filter label to be "Text" instead of "Name"
- Fixed a bug that was causing Monitor to be null for some targets
- Removed sorting from the Monitor column since it was broken and will be a little tricky to implement

@spadgett In addition to the above changes, I also
- Started using React context for the services watcher and service monitors watcher. This fixes a problem where the watchers would be destroyed and recreated when navigating between the list page and details page and also sometimes when scrolling through the list. This change noticeably improve performance.
- Memoized the `Health` component, since it only has 2 states (`up` and `down`)

lihongyan1 commented:
/label qe-approved

spadgett reviewed (approved):
/lgtm

kyoto commented:
Rebased. Could you tag again please @spadgett?

kyoto commented:
/retest

spadgett commented on frontend/public/components/monitoring/targets.tsx:
```diff
@@ -0,0 +1,375 @@
+    monitors,
+    ({ spec }) =>
+      service &&
+      _.isMatch(service.metadata.labels, spec.selector?.matchLabels) &&
```
I missed this previously, but what if the service uses `matchExpressions`?

I'd use `LabelSelector` from label-selector.js to handle `matchExpressions`.

kyoto commented:
Thanks @spadgett. I've updated the PR to use `LabelSelector`.

spadgett reviewed (approved):
/lgtm
